Description: Do not remove duplicated data from status output
Required for backwards compatibility with older releases.
Author: Brett Holman
Origin: other
Last-Update: 2023-10-31
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
--- a/cloudinit/cmd/status.py
+++ b/cloudinit/cmd/status.py
@@ -179,6 +179,9 @@ def print_status(args, details: StatusDe
         "last_update": details.last_update,
         **details.v1,
     }
+    details_dict["schemas"] = {"1": deepcopy(details_dict)}
+    details_dict["_schema_version"] = "1"
+
     if args.format == "tabular":
         prefix = ""
 
--- a/tests/unittests/cmd/test_status.py
+++ b/tests/unittests/cmd/test_status.py
@@ -507,6 +507,7 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr
                 dedent(
                     """\
                    ---
+                   _schema_version: '1'
                    boot_status_code: enabled-by-kernel-command-line
                    datasource: ''
                    detail: 'Running in stage: init'
@@ -520,6 +521,23 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr
                        start: 123.45
                    last_update: Thu, 01 Jan 1970 00:02:04 +0000
                    recoverable_errors: {}
+                   schemas:
+                       '1':
+                           boot_status_code: enabled-by-kernel-command-line
+                           datasource: ''
+                           detail: 'Running in stage: init'
+                           errors: []
+                           extended_status: running
+                           init:
+                               finished: null
+                               start: 124.456
+                           init-local:
+                               finished: 123.46
+                               start: 123.45
+                           last_update: Thu, 01 Jan 1970 00:02:04 +0000
+                           recoverable_errors: {}
+                           stage: init
+                           status: running
                    stage: init
                    status: running
                    ...
@@ -552,6 +570,27 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr
                     "init-local": {"finished": 123.46, "start": 123.45},
                     "last_update": "Thu, 01 Jan 1970 00:02:04 +0000",
                     "recoverable_errors": {},
+                    "_schema_version": "1",
+                    "schemas": {
+                        "1": {
+                            "boot_status_code": (
+                                "enabled-by-kernel-command-line"
+                            ),
+                            "datasource": "",
+                            "detail": "Running in stage: init",
+                            "errors": [],
+                            "extended_status": "running",
+                            "init": {"finished": None, "start": 124.456},
+                            "init-local": {
+                                "finished": 123.46,
+                                "start": 123.45,
+                            },
+                            "last_update": "Thu, 01 Jan 1970 00:02:04 +0000",
+                            "recoverable_errors": {},
+                            "stage": "init",
+                            "status": "running",
+                        }
+                    },
                     "stage": "init",
                 },
                 id="running_json_format",
@@ -583,6 +622,7 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr
                 MyArgs(long=False, wait=False, format="json"),
                 1,
                 {
+                    "_schema_version": "1",
                     "boot_status_code": "enabled-by-kernel-command-line",
                     "datasource": "nocloud",
                     "detail": (
@@ -604,6 +644,32 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr
                     },
                     "last_update": "Thu, 01 Jan 1970 00:02:05 +0000",
                     "recoverable_errors": {},
+                    "schemas": {
+                        "1": {
+                            "boot_status_code": (
+                                "enabled-by-kernel-command-line"
+                            ),
+                            "datasource": "nocloud",
+                            "detail": "DataSourceNoCloud "
+                            "[seed=/var/.../seed/nocloud-net][dsmode=net]",
+                            "errors": ["error1", "error2", "error3"],
+                            "extended_status": "error - running",
+                            "init": {
+                                "errors": ["error1"],
+                                "finished": 125.678,
+                                "start": 124.567,
+                            },
+                            "init-local": {
+                                "errors": ["error2", "error3"],
+                                "finished": 123.46,
+                                "start": 123.45,
+                            },
+                            "last_update": "Thu, 01 Jan 1970 00:02:05 +0000",
+                            "recoverable_errors": {},
+                            "stage": None,
+                            "status": "error",
+                        }
+                    },
                     "stage": None,
                 },
                 id="running_json_format_with_errors",
@@ -666,6 +732,7 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr
                 MyArgs(long=False, wait=False, format="json"),
                 2,
                 {
+                    "_schema_version": "1",
                     "boot_status_code": "enabled-by-kernel-command-line",
                     "datasource": "nocloud",
                     "detail": (
@@ -725,6 +792,91 @@ PATH=/usr/local/sbin:/usr/local/bin:/usr
                             "don't try to open the hatch or we'll all be soup"
                         ],
                     },
+                    "schemas": {
+                        "1": {
+                            "boot_status_code": (
+                                "enabled-by-kernel-command-line"
+                            ),
+                            "datasource": "nocloud",
+                            "detail": "DataSourceNoCloud "
+                            "[seed=/var/.../seed/nocloud-net][dsmode=net]",
+                            "errors": [],
+                            "extended_status": "degraded done",
+                            "init": {
+                                "errors": [],
+                                "finished": 125.678,
+                                "recoverable_errors": {
+                                    "WARNINGS": [
+                                        "the prime "
+                                        "omega "
+                                        "transfuser "
+                                        "borkeded!"
+                                    ]
+                                },
+                                "start": 124.567,
+                            },
+                            "init-local": {
+                                "errors": [],
+                                "finished": 123.46,
+                                "recoverable_errors": {
+                                    "ERROR": [
+                                        "the ion "
+                                        "field "
+                                        "reactor "
+                                        "just "
+                                        "transmutated"
+                                    ]
+                                },
+                                "start": 123.45,
+                            },
+                            "last_update": "Thu, 01 Jan 1970 00:02:08 +0000",
+                            "modules-config": {
+                                "errors": [],
+                                "finished": 126.678,
+                                "recoverable_errors": {
+                                    "CRITICAL": ["Power lost! Prepare to"]
+                                },
+                                "start": 125.567,
+                            },
+                            "modules-final": {
+                                "errors": [],
+                                "finished": 128.678,
+                                "recoverable_errors": {
+                                    "DEPRECATED": [
+                                        "don't "
+                                        "try "
+                                        "to "
+                                        "open "
+                                        "the "
+                                        "hatch "
+                                        "or "
+                                        "we'll "
+                                        "all "
+                                        "be "
+                                        "soup"
+                                    ]
+                                },
+                                "start": 127.567,
+                            },
+                            "recoverable_errors": {
+                                "CRITICAL": ["Power lost! Prepare to"],
+                                "DEPRECATED": [
+                                    "don't try to open "
+                                    "the hatch or we'll "
+                                    "all be soup"
+                                ],
+                                "ERROR": [
+                                    "the ion field reactor "
+                                    "just transmutated"
+                                ],
+                                "WARNINGS": [
+                                    "the prime omega transfuser borkeded!"
+                                ],
+                            },
+                            "stage": None,
+                            "status": "done",
+                        }
+                    },
                     "stage": None,
                 },
                 id="running_json_format_with_recoverable_errors",
